pkg/cloud/amazon/s3_storage.go

```diff
@@ -810,41 +850,38 @@ func (s *s3Storage) List(ctx context.Context, prefix, delim string, fn cloud.Lis
 		return err
 	}
 
-	var fnErr error
-	pageFn := func(page *s3.ListObjectsOutput, lastPage bool) bool {
-		for _, x := range page.CommonPrefixes {
-			if fnErr = fn(strings.TrimPrefix(*x.Prefix, dest)); fnErr != nil {
-				return false
-			}
-		}
-		for _, fileObject := range page.Contents {
-			if fnErr = fn(strings.TrimPrefix(*fileObject.Key, dest)); fnErr != nil {
-				return false
-			}
-		}
-
-		return true
-	}
-
-	var s3Input *s3.ListObjectsInput
+	var s3Input *s3.ListObjectsV2Input
 	// Add an environment variable toggle for s3 storage to list prefixes with a
 	// paging marker that's the prefix with an additional /. This allows certain
 	// s3 clones which return s3://<prefix>/ as the first result of listing
 	// s3://<prefix> to exclude that result.
 	if envutil.EnvOrDefaultBool("COCKROACH_S3_LIST_WITH_PREFIX_SLASH_MARKER", false) {
-		s3Input = &s3.ListObjectsInput{Bucket: s.bucket, Prefix: aws.String(dest), Delimiter: nilIfEmpty(delim), Marker: aws.String(dest + "/")}
+		s3Input = &s3.ListObjectsV2Input{Bucket: s.bucket, Prefix: aws.String(dest), Delimiter: nilIfEmpty(delim), ContinuationToken: aws.String(dest + "/")}
```

Comment:
It looks like we don't have any tests for this behaviour.  I wonder if this works the same since the ContinuationToken documentation says that the argument is opaque and not an actual key.